### PR TITLE
[7.7.x] RHBA-71 - Allow forward operation for Business Administrators

### PR DESCRIPTION
**Pull request by cristianonicolai**

### CODE REVIEW

**Comment by cristianonicolai** [comment]
@mswiderski had to remove the explicit requirement for user to be in the potential owner list, as a BA that is not true. Also, in this case, previous potential owner is kept. Sounds like a valid approach?

**Comment by mswiderski** [comment]
@cristianonicolai isn't that now we actually open the forward operation for other users too - sine the check of being a potential owner is removed? Forward has some restriction that it can only be done when task is assigned to individuals and not groups and with this change it might be now possible to forward task that is group assigned

> task‟s potential owners, actual owner or business administrator can forward an active task to
another person or a set of people, replacing himself by those people in the list of potential owners.

but to be honest not sure how would we interpret it - if we follow it strictly this is what we currently have so user who performs the operation must be one of the potential owners but does it really make sense for business admins... not so sure about it.

Can we add a test where task is assigned to group and then business admin tries to forward it? According to spec it should not be allowed so if that won't be allowed with this change then we are good to go but if the task will be forwarded then we break the spec in terms of behaviour and then I don't think we should do this (at least not that way).

**Comment by cristianonicolai** [comment]
@mswiderski agree, it leaves open to interpretation. I could prevent forwarding from a group, but had to include another property to _OperationCommand_ just for that, if you have any better idea...

**Comment by mswiderski** [comment]
@cristianonicolai maybe we could use similar approach as Fail operation that and (ab)use the before/after event handling https://github.com/cristianonicolai/jbpm/blob/c598eb1a65cc71b3432a6628b361975e4204b5fe/jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/internals/lifecycle/MVELLifeCycleManager.java#L359

wdyt?

**Comment by cristianonicolai** [comment]
@mswiderski another point is that I don't think it makes sense for me to try to completely duplicate these rules in the UI to prevent a user from having a button enabled to forward a task. That leads to another issue, I can't differentiate an task not found from a permission exception error in the UI because it gets wrapped in here: https://github.com/kiegroup/jbpm/blob/master/jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/UserTaskServiceImpl.java#L372
I'm wondering if we could expose it in REST interface to better handle it in the UI, although iI understand we might want to leave that hidden from the REST endpoint consumer.

**Comment by cristianonicolai** [comment]
@mswiderski sure, let me try that.

File: jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/internals/lifecycle/MVELLifeCycleManager.java [hunk 1]
```diff
@@ -351,7 +351,12 @@ public void taskOperation(final Operation operation, final long taskId, final St
                     break;
                 }
                 case Forward: {
-                	taskEventSupport.fireBeforeTaskForwarded(task, context);
+                    taskEventSupport.fireBeforeTaskForwarded(task, context);
+                    if (task.getPeopleAssignments().getPotentialOwners().stream().anyMatch(oe -> oe instanceof Group)) {
```
[the remark this answers is not in the document]
Comment (mswiderski) [reply jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/internals/lifecycle/MVELLifeCycleManager.java]: yes, pretty much like this - though maybe for error message we can simply say forward not allowed for task with group assignment?

[the remark this answers is not in the document]
Comment (cristianonicolai) [reply jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/internals/lifecycle/MVELLifeCycleManager.java]: ok, done

**Comment by cristianonicolai** [comment]
@mswiderski when you have a chance, this should be ready for review with latest changes.

**Comment by cristianonicolai** [comment]
Jenkins execute full downstream build

**Pull request merged**